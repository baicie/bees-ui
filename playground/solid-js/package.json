{
  "name": "solid-js",
  "private": true,
  "version": "0.0.0",
  "type": "module",
  "scripts": {
    "start": "vite",
    "build-play": "tsc && vite build",
    "preview": "vite preview"
  },
  "dependencies": {
    "solid-js": "^1.8.7"
  },
  "devDependencies": {
    "typescript": "^5.3.3",
    "vite": "^5.0.12",
    "vite-plugin-solid": "^2.9.1"
  }
}