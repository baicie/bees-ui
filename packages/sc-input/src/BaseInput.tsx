import clsx from 'clsx';
import { children, Show, splitProps } from 'solid-js';
import { Dynamic } from 'solid-js/web';

import type { BaseInputProps } from './interface';
import { cloneChildren, hasAddon, hasPrefixSuffix } from './utils/commonUtils';

const BaseInput = (props: BaseInputProps) => {
  const [local] = splitProps(props, [
    'children',
    'prefixCls',
    'prefix',
    'suffix',
    'addonBefore',
    'addonAfter',
    'className',
    'style',
    'disabled',
    'readOnly',
    'focused',
    'triggerFocus',
    'allowClear',
    'value',
    'handleReset',
    'hidden',
    'classes',
    'classNames',
    'dataAttrs',
    'styles',
    'components',
    'onClear',
  ]);

  const AffixWrapperComponent = local.components?.affixWrapper || 'span';
  const GroupWrapperComponent = local.components?.groupWrapper || 'span';
  const WrapperComponent = local.components?.wrapper || 'span';
  const GroupAddonComponent = local.components?.groupAddon || 'span';

  const inputElement = children(() => local.children);

  const containerRef: HTMLDivElement | null = null;

  const onInputClick = (e: MouseEvent) => {
    if (containerRef?.contains(e.target as Element)) {
      local.triggerFocus?.();
    }
  };

  const hasAffix = hasPrefixSuffix(local);

  let element = cloneChildren(inputElement, {
    value: local.value,
    class: clsx(!hasAffix && local.classNames?.variant) || null,
  });

  // ================== Prefix & Suffix ================== //
  if (hasAffix) {
    let clearIcon = null;
    if (local.allowClear) {
      const needClear = !local.disabled && !local.readOnly && local.value;
      const clearIconCls = `${local.prefixCls}-clear-icon`;
      const iconNode =
        typeof local.allowClear === 'object' && local.allowClear?.clearIcon
          ? local.allowClear.clearIcon
          : 'âœ–';

      clearIcon = (
        <span
          onClick={(e) => {
            local.handleReset?.(e);
            local.onClear?.();
          }}
          onMouseDown={(e) => e.preventDefault()}
          class={clsx(clearIconCls, {
            [`${clearIconCls}-hidden`]: !needClear,
            [`${clearIconCls}-has-suffix`]: !!local.suffix,
          })}
          role="button"
          tabIndex={-1}
        >
          {iconNode}
        </span>
      );
    }

    element = (
      <Dynamic
        component={AffixWrapperComponent}
        class={clsx(
          `${local.prefixCls}-affix-wrapper`,
          {
            [`${local.prefixCls}-disabled`]: local.disabled,
            [`${local.prefixCls}-focused`]: local.focused,
            [`${local.prefixCls}-readonly`]: local.readOnly,
          },
          local.classes?.affixWrapper,
          local.classNames?.affixWrapper,
        )}
        style={local.styles?.affixWrapper}
        onClick={onInputClick}
        ref={containerRef}
        {...local.dataAttrs?.affixWrapper}
      >
        <Show when={local.prefix}>
          <span
            class={clsx(`${local.prefixCls}-prefix`, local.classNames?.prefix)}
            style={local.styles?.prefix}
          >
            {local.prefix}
          </span>
        </Show>

        {element}

        <Show when={local.suffix || local.allowClear}>
          <span
            class={clsx(`${local.prefixCls}-suffix`, local.classNames?.suffix)}
            style={local.styles?.suffix}
          >
            {clearIcon}
            {local.suffix}
          </span>
        </Show>
      </Dynamic>
    );
  }

  // ================== Addon ================== //
  if (hasAddon(local)) {
    const wrapperCls = `${local.prefixCls}-group`;
    const mergedWrapperClassName = clsx(
      `${local.prefixCls}-wrapper`,
      wrapperCls,
      local.classes?.wrapper,
      local.classNames?.wrapper,
    );

    element = (
      <Dynamic
        component={GroupWrapperComponent}
        class={clsx(
          `${wrapperCls}-wrapper`,
          { [`${wrapperCls}-disabled`]: local.disabled },
          local.classes?.group,
          local.classNames?.groupWrapper,
        )}
      >
        <Dynamic component={WrapperComponent} class={mergedWrapperClassName}>
          {local.addonBefore && (
            <Dynamic component={GroupAddonComponent} class={`${wrapperCls}-addon`}>
              {local.addonBefore}
            </Dynamic>
          )}
          {element}
          {local.addonAfter && (
            <Dynamic component={GroupAddonComponent} class={`${wrapperCls}-addon`}>
              {local.addonAfter}
            </Dynamic>
          )}
        </Dynamic>
      </Dynamic>
    );
  }

  return element;
};

export default BaseInput;
