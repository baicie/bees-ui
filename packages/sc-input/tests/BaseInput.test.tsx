import { fireEvent, render } from '@solidjs/testing-library';
import { createSignal } from 'solid-js';

import BaseInput from '../src/BaseInput';

describe('BaseInput', () => {
  it('should render perfectly', () => {
    const { container } = render(() => (
      <BaseInput prefixCls="rc-input">
        <input />
      </BaseInput>
    ));
    expect(container).toMatchSnapshot();
  });

  it('prefix and suffix should render properly', () => {
    const { container } = render(() => (
      <div>
        <BaseInput prefixCls="rc-input" prefix="prefix">
          <input />
        </BaseInput>
        <br />
        <br />
        <BaseInput prefixCls="rc-input" suffix="suffix">
          <input />
        </BaseInput>
      </div>
    ));
    expect(container).toMatchSnapshot();
  });

  it('addon should render properly', () => {
    const { container } = render(() => (
      <div>
        <BaseInput prefixCls="rc-input" addonBefore="addonBefore">
          <input />
        </BaseInput>
        <br />
        <br />
        <BaseInput prefixCls="rc-input" addonAfter="addonAfter">
          <input />
        </BaseInput>
      </div>
    ));
    expect(container).toMatchSnapshot();
  });

  it('allowClear should work', () => {
    const onChange = vitest.fn();
    const onBlur = vitest.fn();
    const onFocus = vitest.fn();

    const Demo = () => {
      const [value, setValue] = createSignal('');

      const handleReset = () => {
        setValue('');
      };

      const handleChange = (e: Event) => {
        onChange();
        setValue((e.target as HTMLInputElement).value);
      };

      return (
        <BaseInput
          prefixCls="rc-input"
          allowClear={{ clearIcon: '✖' }}
          value={value()}
          handleReset={handleReset}
        >
          <input onChange={handleChange} onBlur={onBlur} onFocus={onFocus} />
        </BaseInput>
      );
    };

    const { container } = render(() => <Demo />);

    const inputEl = container.querySelector('input');
    fireEvent.focus(inputEl!);
    expect(onFocus).toHaveBeenCalledTimes(1);

    fireEvent.change(inputEl!, { target: { value: 'some text' } });
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(inputEl!.value).toBe('some text');

    const clearIcon = container.querySelector('.rc-input-clear-icon');
    fireEvent.mouseDown(clearIcon!);
    fireEvent.click(clearIcon!);
    fireEvent.mouseUp(clearIcon!);
    expect(onBlur).not.toHaveBeenCalled();
    expect(onChange).toHaveBeenCalledTimes(1);
    expect(inputEl!.value).toBe('');
  });

  it('should display clearIcon correctly', () => {
    const { container, unmount } = render(() => (
      <BaseInput prefixCls="rc-input" allowClear>
        <input />
      </BaseInput>
    ));
    const clearIcon = container.querySelector('.rc-input-clear-icon');
    expect(clearIcon?.innerHTML).toBe('✖');

    unmount();

    render(() => (
      <BaseInput prefixCls="rc-input" allowClear={{ clearIcon: 'clear' }}>
        <input />
      </BaseInput>
    ));
    expect(clearIcon?.innerHTML).toBe('clear');
  });

  it('should focus input when click prefix', () => {
    let inputRef: HTMLInputElement | null = null;
    const { container } = render(() => (
      <BaseInput prefixCls="rc-input" triggerFocus={() => inputRef?.focus()} prefix="$">
        <input ref={(el) => (inputRef = el)} />
      </BaseInput>
    ));
    fireEvent.click(container.querySelector('.rc-input-affix-wrapper')!);
    expect(document.activeElement).toBe(inputRef);
  });

  it('should support data-*', () => {
    const { container } = render(() => (
      <BaseInput
        prefixCls="rc-input"
        prefix="prefix"
        dataAttrs={{
          affixWrapper: {
            'data-test': 'test',
          },
        }}
      >
        <input />
      </BaseInput>
    ));
    expect(container.querySelector('.rc-input-affix-wrapper')?.getAttribute('data-test')).toBe(
      'test',
    );
  });

  it('should apply className to inputElement', () => {
    const { container } = render(() => (
      <BaseInput prefixCls="rc-input" className="test-base">
        <input class="test" />
      </BaseInput>
    ));
    expect(container.querySelector('.test-base')).toBeTruthy();
    expect(container.querySelector('.test')).toBeTruthy();
  });

  it('should not pass className to inputElement when has addon', () => {
    const { container } = render(() => (
      <BaseInput prefixCls="rc-input" className="test-base" addonBefore="addon">
        <input class="test" />
      </BaseInput>
    ));
    expect(container.querySelector('input')?.classList.contains('test')).toBeTruthy();
    expect(container.querySelector('input')?.classList.contains('test-base')).toBeFalsy();
  });

  it('should correct render with prefix and addon', () => {
    const { container } = render(() => (
      <BaseInput prefixCls="rc-input" prefix="prefix" addonBefore="addon">
        <input />
      </BaseInput>
    ));
    expect(container).toMatchSnapshot();
  });

  it('support use div as basic element', () => {
    const { container } = render(() => (
      <BaseInput
        prefixCls="rc-input"
        prefix="prefix"
        addonBefore="addon"
        components={{
          affixWrapper: 'div',
          groupWrapper: 'div',
          wrapper: 'div',
          groupAddon: 'div',
        }}
      >
        <input />
      </BaseInput>
    ));
    expect(container).toMatchSnapshot();
  });

  it('styles should work', () => {
    const container = render(() => (
      <BaseInput
        prefixCls="rc-input"
        prefix="prefix"
        addonBefore="addon"
        styles={{
          affixWrapper: {
            color: 'red',
          },
          prefix: {
            color: 'blue',
          },
        }}
      >
        <input />
      </BaseInput>
    ));

    expect(
      container.container.querySelector<HTMLSpanElement>('.rc-input-affix-wrapper')?.style.color,
    ).toBe('red');
    expect(
      container.container.querySelector<HTMLSpanElement>('.rc-input-prefix')?.style.color,
    ).toBe('blue');
  });

  it('with addon and disabled', () => {
    const { container } = render(() => (
      <BaseInput prefixCls="rc-input" addonBefore="addon" disabled>
        <input />
      </BaseInput>
    ));

    expect(container.firstChild).toHaveClass('rc-input-group-wrapper-disabled');
  });

  it('variant cls', () => {
    const { container, unmount } = render(() => (
      <BaseInput prefixCls="rc-input" prefix="$" classNames={{ variant: 'test-variant' }} disabled>
        <input />
      </BaseInput>
    ));

    expect(container.querySelector('.rc-input-affix-wrapper')).toHaveClass('test-variant');
    expect(container.querySelector('input')).not.toHaveClass('test-variant');

    unmount();

    render(() => (
      <BaseInput prefixCls="rc-input" classNames={{ variant: 'test-variant' }} disabled>
        <input />
      </BaseInput>
    ));

    expect(container.querySelector('.rc-input-affix-wrapper')).toBeFalsy();
    expect(container.querySelector('input')).toHaveClass('test-variant');
  });

  describe('ref', () => {
    it('prefix', () => {
      let holderRef: any = null;
      const { container } = render(() => (
        <BaseInput prefixCls="rc-input" prefix="prefix" ref={(el) => (holderRef = el)}>
          <input />
        </BaseInput>
      ));
      expect(holderRef?.nativeElement).toBe(container.querySelector('.rc-input-affix-wrapper'));
    });

    it('addon', () => {
      let holderRef: any = null;
      const { container } = render(() => (
        <BaseInput prefixCls="rc-input" addonAfter="after" ref={(el) => (holderRef = el)}>
          <input />
        </BaseInput>
      ));

      expect(holderRef?.nativeElement).toBe(container.querySelector('.rc-input-group-wrapper'));
    });

    it('mix', () => {
      let holderRef: any = null;
      const { container } = render(() => (
        <BaseInput
          prefixCls="rc-input"
          suffix="suffix"
          addonAfter="after"
          ref={(el) => (holderRef = el)}
        >
          <input />
        </BaseInput>
      ));

      expect(holderRef?.nativeElement).toBe(container.querySelector('.rc-input-group-wrapper'));
    });

    it('support onClear', () => {
      const onClear = vitest.fn();
      const { container } = render(() => (
        <BaseInput prefixCls="rc-input" onClear={onClear} allowClear>
          <input value="test" />
        </BaseInput>
      ));
      fireEvent.click(container.querySelector<HTMLSpanElement>('.rc-input-clear-icon')!);
      expect(onClear).toHaveBeenCalled();
    });
  });
});
