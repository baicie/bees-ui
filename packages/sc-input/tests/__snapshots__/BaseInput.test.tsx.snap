// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BaseInput > addon should render properly 1`] = `
<div>
  <div>
    <span
      class="rc-input-group-wrapper"
    >
      <span
        class="rc-input-wrapper rc-input-group"
      >
        <span
          class="rc-input-group-addon"
        >
          addonBefore
        </span>
        <input />
      </span>
    </span>
    <br />
    <br />
    <span
      class="rc-input-group-wrapper"
    >
      <span
        class="rc-input-wrapper rc-input-group"
      >
        <input />
        <span
          class="rc-input-group-addon"
        >
          addonAfter
        </span>
      </span>
    </span>
  </div>
</div>
`;

exports[`BaseInput > prefix and suffix should render properly 1`] = `
<div>
  <div>
    <span
      class="rc-input-affix-wrapper"
    >
      <span
        class="rc-input-prefix"
      >
        prefix
      </span>
      <input />
    </span>
    <br />
    <br />
    <span
      class="rc-input-affix-wrapper"
    >
      <input />
      <span
        class="rc-input-suffix"
      >
        
        suffix
      </span>
    </span>
  </div>
</div>
`;

exports[`BaseInput > should correct render with prefix and addon 1`] = `
<div>
  <span
    class="rc-input-group-wrapper"
  >
    <span
      class="rc-input-wrapper rc-input-group"
    >
      <span
        class="rc-input-group-addon"
      >
        addon
      </span>
      <span
        class="rc-input-affix-wrapper"
      >
        <span
          class="rc-input-prefix"
        >
          prefix
        </span>
        <input />
      </span>
    </span>
  </span>
</div>
`;

exports[`BaseInput > should render perfectly 1`] = `
<div>
  <input />
</div>
`;

exports[`BaseInput > support use div as basic element 1`] = `
<div>
  <div
    class="rc-input-group-wrapper"
  >
    <div
      class="rc-input-wrapper rc-input-group"
    >
      <div
        class="rc-input-group-addon"
      >
        addon
      </div>
      <div
        class="rc-input-affix-wrapper"
      >
        <span
          class="rc-input-prefix"
        >
          prefix
        </span>
        <input />
      </div>
    </div>
  </div>
</div>
`;
